nixos/modules/services/web-apps/lemmy.nix

```
@@ -0,0 +1,142 @@
+{ lib, pkgs, config, ... }:
+with lib;
+let
+  cfg = config.services.lemmy;
+  settingsFormat = pkgs.formats.json { };
+in
+{
+  options.services.lemmy = {
+    enable = mkEnableOption "Enable lemmy server";
+
+    config = mkOption {
+      type = types.submodule {
+        freeformType = settingsFormat.type;
+        options.database = {
+          password = mkOption {
+            type = types.str;
+            description = "password to connect to postgres";
+          };
+        };
+        options.hostname = mkOption {
+          type = types.str;
+          default = null;
+          description = ''the domain name of your instance (eg "lemmy.ml")'';
+        };
+        options.port = mkOption {
+          type = types.port;
+          default = 8536;
+          description = "port where lemmy should listen for incoming requests";
+        };
+        # TODO: this is a secret, should not be in nix-store
+        options.jwt_secret = mkOption {
+          type = types.str;
+          description = "json web token for authorization between server and client";
+        };
+        options.federation = {
+          enabled = mkEnableOption "whether to enable activitypub federation.";
+        };
+        options.captcha = {
+          enabled = mkOption {
+            type = types.bool;
+            default = true;
+            description = "Enable Captcha";
+          };
+          difficulty = mkOption {
+            type = types.str; # TODO: Types should be "easy", "medium", "hard". Define an enum.
+            default = "medium";
+            description = "The difficultly of the captcha to solve";
+          };
+        };
+      };
+      default = { };
+      description = "Lemmy configuration";
+    };
+
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.lemmy.settings = {
```

```suggestion
    services.lemmy.settings = mapAttrs (name: mkDefault) {
```

Avoid RSI . Type `mkDefault` once only. 😉

---

The functional programmer in me is mad I didn't think of doing this.
Super cool!